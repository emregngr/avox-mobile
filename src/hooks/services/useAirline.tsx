import { useQuery } from '@tanstack/react-query'
import { useEffect, useMemo, useState } from 'react'

import { getAllAirlines } from '@/services/airlineService'
import useLocaleStore from '@/store/locale'
import type { Airline } from '@/types/feature/airline'
import { useDebounce } from '@/utils/common/useDebounce'
import { parseFilterRange } from '@/utils/feature/parseFilterRange'

const ITEMS_PER_PAGE = 20

export function useAirline() {
  const { selectedLocale } = useLocaleStore()
  const [searchTerm, setSearchTerm] = useState<string>('')
  const [filters, setFilters] = useState<any>({})
  const [page, setPage] = useState<number>(1)
  const [isSearchLoading, setIsSearchLoading] = useState<boolean>(false)
  const debouncedSearch = useDebounce<string>(searchTerm, 500)

  const {
    data: allAirlinesData,
    error,
    isLoading,
  } = useQuery({
    queryFn: () => getAllAirlines(selectedLocale),
    queryKey: ['allAirlines', selectedLocale],
    staleTime: 5 * 60 * 1000,
  })

  const applyFilters = (airlines: Airline[]): Airline[] => {
    if (!airlines) return []

    return airlines.filter(airline => {
      if (!airline) return false

      if (filters?.region) {
        const airlineRegion = airline?.operations?.region?.toLowerCase()
        const filterRegion = filters?.region?.toLowerCase()

        if (airlineRegion !== filterRegion) {
          return false
        }
      }

      if (filters?.businessType) {
        const businessType = airline?.operations?.businessType?.toLowerCase()
        const filterBusinessType = filters?.businessType?.toLowerCase()

        if (businessType !== filterBusinessType) {
          return false
        }
      }

      if (filters?.businessModel) {
        const businessModel = airline?.operations?.businessModel?.toLowerCase()
        const filterBusinessModel = filters?.businessModel?.toLowerCase()

        if (businessModel !== filterBusinessModel) {
          return false
        }
      }

      if (filters?.alliance) {
        const alliance = airline?.operations?.alliance?.toLowerCase()
        const filterAlliance = filters?.alliance?.toLowerCase()

        if (alliance !== filterAlliance) {
          return false
        }
      }

      if (filters?.minSkytraxRating && airline?.operations?.skytraxRating) {
        if (airline?.operations?.skytraxRating < filters?.minSkytraxRating) {
          return false
        }
      }

      const airlineRangeChecks: {
        filterKey: keyof typeof filters
        isYear?: boolean
        jsonPath: (a: Airline) => number | undefined | null
      }[] = [
        {
          filterKey: 'foundingYear',
          isYear: true,
          jsonPath: a =>
            a.companyInfo?.foundingYear
              ? parseInt(String(a?.companyInfo?.foundingYear), 10)
              : undefined,
        },
        { filterKey: 'passengerCapacity', jsonPath: a => a?.companyInfo?.passengerCapacity },
        { filterKey: 'employeeCount', jsonPath: a => a?.companyInfo?.employeeCount },
        { filterKey: 'destinationCount', jsonPath: a => a?.network?.destinationCount },
        { filterKey: 'destinationCountries', jsonPath: a => a?.network?.destinationCountries },
        { filterKey: 'domesticConnections', jsonPath: a => a?.network?.domesticConnections },
        {
          filterKey: 'internationalConnections',
          jsonPath: a => a?.network?.internationalConnections,
        },
        { filterKey: 'totalAirplane', jsonPath: a => a?.fleet?.totalAirplane },
        { filterKey: 'averageAgeYears', jsonPath: a => a?.fleet?.averageAgeYears },
        { filterKey: 'airplaneTypeCount', jsonPath: a => a?.fleet?.airplaneTypeCount },
      ]

      for (const check of airlineRangeChecks) {
        const filterValueStr = filters?.[check.filterKey] as string
        if (filterValueStr) {
          const parsed = parseFilterRange(filterValueStr)
          if (parsed) {
            let itemValue = check?.jsonPath(airline)
            if (itemValue === undefined || itemValue === null || isNaN(Number(itemValue))) {
              return false
            }
            itemValue = Number(itemValue)
            if (itemValue < parsed.min || itemValue > parsed.max) {
              return false
            }
          }
        }
      }
      return true
    })
  }

  const filteredAirlines = useMemo(() => {
    if (!allAirlinesData) return []

    let filtered = allAirlinesData

    if (debouncedSearch.trim()) {
      const searchLower = debouncedSearch.toLowerCase().trim()

      const exactMatches = filtered.filter(
        airline =>
          airline.iataCode?.toLowerCase() === searchLower ||
          airline.icaoCode?.toLowerCase() === searchLower,
      )

      if (exactMatches.length > 0) {
        filtered = exactMatches
      } else {
        filtered = filtered.filter(airline => {
          const name = airline?.name?.toLowerCase() || ''
          const city = airline?.operations?.hub?.city?.toLowerCase() || ''
          const country = airline?.operations?.country?.toLowerCase() || ''
          const iata = airline?.iataCode?.toLowerCase() || ''
          const icao = airline?.icaoCode?.toLowerCase() || ''

          return (
            name.includes(searchLower) ||
            city.includes(searchLower) ||
            country.includes(searchLower) ||
            iata.includes(searchLower) ||
            icao.includes(searchLower)
          )
        })
      }
    }

    return applyFilters(filtered)
  }, [allAirlinesData, debouncedSearch, filters])

  const paginatedAirlines = useMemo(() => {
    if (!filteredAirlines) return []
    return filteredAirlines.slice(0, page * ITEMS_PER_PAGE)
  }, [filteredAirlines, page])

  const hasMore = paginatedAirlines.length < filteredAirlines.length

  const loadMore = () => {
    if (hasMore && !isLoading) {
      setPage(prev => prev + 1)
    }
  }

  useEffect(() => {
    if (searchTerm !== debouncedSearch) {
      setIsSearchLoading(true)
    } else {
      setTimeout(() => setIsSearchLoading(false), 300)
    }
  }, [searchTerm, debouncedSearch])

  useEffect(() => {
    setPage(1)
  }, [debouncedSearch, filters])

  return {
    error,
    filteredCount: filteredAirlines?.length,
    filters,
    hasMore,
    isLoading,
    isSearchLoading,
    loadMore,
    paginatedAirlines,
    searchTerm,
    setFilters,
    setSearchTerm,
  }
}
