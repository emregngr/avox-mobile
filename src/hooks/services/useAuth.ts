import { getApp } from '@react-native-firebase/app'
import type { FirebaseAuthTypes } from '@react-native-firebase/auth'
import { getAuth, getIdToken, onAuthStateChanged } from '@react-native-firebase/auth'
import type { QueryClient } from '@tanstack/react-query'
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { router } from 'expo-router'
import { Alert } from 'react-native'
import Toast from 'react-native-toast-message'

import { getLocale } from '@/locales/i18next'
import {
  handleLogout,
  sendPasswordResetLink,
  signInWithApple,
  signInWithEmail,
  signInWithGoogle,
  signUpWithEmail,
} from '@/services/authService'
import { login, logout, register, social } from '@/store/auth'
import useThemeStore from '@/store/theme'
import type { AuthCredentials, LoginCredentials, RegisterCredentials } from '@/types/feature/auth'
import { Logger } from '@/utils/common/logger'

const app = getApp()
const auth = getAuth(app)

const handleAuthSuccess = async <T extends AuthCredentials>(
  userCredential: FirebaseAuthTypes.UserCredential,
  authFunction: (data: T & { token: string }) => Promise<void>,
  authData: T,
  queryClient: QueryClient,
) => {
  try {
    const token = await getIdToken(userCredential.user, true)
    await authFunction({ ...authData, token })
    queryClient.invalidateQueries({ queryKey: ['user'] })
  } catch (error) {
    Logger.breadcrumb('Failed to handle auth success', 'error', error as Error)
  }
}

export const useUser = () =>
  useQuery({
    queryFn: () =>
      new Promise(resolve => {
        const unsubscribe = onAuthStateChanged(auth, userState => {
          unsubscribe()
          resolve(userState)
        })
      }),
    queryKey: ['user'],
    staleTime: Infinity,
  })

export const useEmailRegister = () => {
  const queryClient = useQueryClient()
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: async ({ email, firstName, lastName, password }: RegisterCredentials) =>
      await signUpWithEmail({ email, firstName, lastName, password }),
    onError: () =>
      Alert.alert(
        getLocale('error'),
        getLocale('emailRegisterFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      ),
    onSuccess: async (userCredential, variables) => {
      await handleAuthSuccess(
        userCredential,
        register,
        {
          email: variables.email,
          firstName: variables.firstName,
          lastName: variables.lastName,
          password: variables.password,
          token: '',
        },
        queryClient,
      )
    },
  })
}

export const useEmailLogin = () => {
  const queryClient = useQueryClient()
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: async ({ email, password }: LoginCredentials) =>
      await signInWithEmail({ email, password }),
    onError: () =>
      Alert.alert(
        getLocale('error'),
        getLocale('emailLoginFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      ),
    onSuccess: async (userCredential, variables) => {
      await handleAuthSuccess(
        userCredential,
        login,
        {
          email: variables.email,
          password: variables.password,
        },
        queryClient,
      )
    },
  })
}

const useSocialLogin = (
  signInMethod: () => Promise<FirebaseAuthTypes.UserCredential>,
  provider: 'google' | 'apple',
) => {
  const queryClient = useQueryClient()
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: async () => await signInMethod(),
    onError: () =>
      Alert.alert(
        getLocale('error'),
        getLocale('socialLoginFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      ),
    onSuccess: async userCredential => {
      await handleAuthSuccess(userCredential, social, { provider }, queryClient)
    },
  })
}

export const useGoogleLogin = () => useSocialLogin(signInWithGoogle, 'google')

export const useAppleLogin = () => useSocialLogin(signInWithApple, 'apple')

export const useLogout = () => {
  const queryClient = useQueryClient()
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: async () => await handleLogout(),
    onError: () => {
      Alert.alert(
        getLocale('error'),
        getLocale('logoutFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      )
    },
    onSuccess: async () => {
      await logout()
      queryClient.setQueryData(['user'], null)
      queryClient.removeQueries()
    },
  })
}

export const useForgotPassword = () => {
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: sendPasswordResetLink,
    onError: () =>
      Alert.alert(
        getLocale('error'),
        getLocale('passwordResetLinkSentMessageFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      ),
    onSuccess: () => {
      Toast.show({
        text1: getLocale('successful'),
        text2: getLocale('passwordResetLinkSentMessage'),
        type: 'success',
      })
      router.back()
    },
  })
}
