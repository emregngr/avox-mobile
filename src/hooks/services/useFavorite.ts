import { getApp } from '@react-native-firebase/app'
import { getAuth } from '@react-native-firebase/auth'
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { useFocusEffect } from 'expo-router'
import { useCallback } from 'react'

import {
  addToFavorites,
  fetchFavoriteDetails,
  fetchFavoriteIds,
  removeFromFavorites,
} from '@/services/favoriteService'
import useLocaleStore from '@/store/locale'
import type { FavoriteItem } from '@/types/feature/favorite'

const app = getApp()
const auth = getAuth(app)

const QUERY_KEYS = {
  favoriteDetails: (userId: string, locale: string) => ['favoriteDetails', userId, locale],
  favorites: (userId: string) => ['favorites', userId],
} as const

const getCurrentUserId = () => auth?.currentUser?.uid as string

export const useFavoriteIds = (refetchOnFocus = false) => {
  const userId = getCurrentUserId()
  const queryClient = useQueryClient()

  const query = useQuery({
    enabled: !!userId,
    gcTime: 5 * 60 * 1000,
    queryFn: fetchFavoriteIds,
    queryKey: QUERY_KEYS?.favorites(userId),
    staleTime: 1 * 60 * 1000,
  })

  useFocusEffect(
    useCallback(() => {
      const getFavorites = async () => {
        if (refetchOnFocus && userId) {
          await queryClient.invalidateQueries({
            queryKey: QUERY_KEYS?.favorites(userId),
          })
        }
      }
      getFavorites()
    }, [refetchOnFocus, userId, queryClient]),
  )

  return query
}

export const useFavoriteDetails = (refetchOnFocus = false) => {
  const { selectedLocale } = useLocaleStore()
  const userId = getCurrentUserId()
  const queryClient = useQueryClient()
  const { data: favoriteIds = [] } = useFavoriteIds(refetchOnFocus)

  const query = useQuery({
    enabled: !!userId && !!selectedLocale,
    queryFn: async () => {
      if (!favoriteIds || favoriteIds?.length === 0) return Promise.resolve([])
      const favoriteDetails = await fetchFavoriteDetails(favoriteIds, selectedLocale)
      return favoriteDetails
    },
    queryKey: QUERY_KEYS?.favoriteDetails(userId, selectedLocale),
    staleTime: 1 * 60 * 1000,
  })

  useFocusEffect(
    useCallback(() => {
      const getFavoriteDetails = async () => {
        if (refetchOnFocus && userId) {
          await queryClient?.invalidateQueries({
            queryKey: QUERY_KEYS?.favoriteDetails(userId, selectedLocale),
          })
        }
      }
      getFavoriteDetails()
    }, [refetchOnFocus, userId, selectedLocale, queryClient, favoriteIds]),
  )

  return query
}

export const useAddFavorite = () => {
  const queryClient = useQueryClient()
  const userId = getCurrentUserId()
  const { selectedLocale } = useLocaleStore()

  return useMutation({
    mutationFn: addToFavorites,

    onError: (err, variables, context) => {
      if (context?.previousFavorites && userId) {
        queryClient?.setQueryData(QUERY_KEYS.favorites(userId), context?.previousFavorites)
      }
    },

    onMutate: async (variables: FavoriteItem) => {
      if (!userId) return

      await queryClient.cancelQueries({
        queryKey: QUERY_KEYS.favorites(userId),
      })

      const previousFavorites = queryClient?.getQueryData(QUERY_KEYS.favorites(userId))

      queryClient.setQueryData(QUERY_KEYS.favorites(userId), (old: any) => {
        if (!old) return [{ id: variables?.id, type: variables?.type }]

        const exists = old.some(
          (item: any) => item?.id === variables?.id && item?.type === variables?.type,
        )
        if (exists) return old
        return [...old, { id: variables?.id, type: variables?.type }]
      })

      return { previousFavorites }
    },

    onSuccess: async () => {
      if (!userId) return

      await queryClient.invalidateQueries({
        queryKey: QUERY_KEYS.favorites(userId),
      })

      await queryClient.invalidateQueries({
        queryKey: QUERY_KEYS.favoriteDetails(userId, selectedLocale),
      })
    },
  })
}

export const useRemoveFavorite = () => {
  const queryClient = useQueryClient()
  const userId = getCurrentUserId()
  const { selectedLocale } = useLocaleStore()

  return useMutation({
    mutationFn: removeFromFavorites,

    onError: (err, variables, context) => {
      if (context?.previousFavorites && userId) {
        queryClient?.setQueryData(QUERY_KEYS.favorites(userId), context?.previousFavorites)
      }
    },

    onMutate: async (variables: FavoriteItem) => {
      if (!userId) return

      await queryClient.cancelQueries({
        queryKey: QUERY_KEYS.favorites(userId),
      })

      const previousFavorites = queryClient.getQueryData(QUERY_KEYS.favorites(userId))

      queryClient.setQueryData(QUERY_KEYS.favorites(userId), (old: any) => {
        if (!old) return []
        return old?.filter(
          (item: any) => !(item?.id === variables?.id && item?.type === variables?.type),
        )
      })

      return { previousFavorites }
    },

    onSuccess: async () => {
      if (!userId) return

      await queryClient.invalidateQueries({
        queryKey: QUERY_KEYS.favorites(userId),
      })

      await queryClient.invalidateQueries({
        queryKey: QUERY_KEYS.favoriteDetails(userId, selectedLocale),
      })
    },
  })
}

export const useIsFavorite = ({ id, type }: FavoriteItem) => {
  const { data: favoriteIds = [] } = useFavoriteIds()

  return favoriteIds?.some((fav: FavoriteItem) => fav?.id === id && fav?.type === type)
}
