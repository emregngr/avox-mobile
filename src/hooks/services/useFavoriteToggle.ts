import { router } from 'expo-router'
import { useCallback, useMemo } from 'react'

import { useAddFavorite, useIsFavorite, useRemoveFavorite } from '@/hooks/services/useFavorite'
import useAuthStore from '@/store/auth'
import type { FavoriteItemType } from '@/types/feature/favorite'
import { Logger } from '@/utils/common/logger'

export function useFavoriteToggle({ id, type }: FavoriteItemType) {
  const { isAuthenticated } = useAuthStore()

  const { isPending: isAddFavoritePending, mutateAsync: addFavoriteMutation } = useAddFavorite()
  const { isPending: isRemoveFavoritePending, mutateAsync: removeFavoriteMutation } =
    useRemoveFavorite()

  const isFavorite = useIsFavorite({ id, type })

  const isPending = useMemo(
    () => isAddFavoritePending || isRemoveFavoritePending,
    [isAddFavoritePending, isRemoveFavoritePending],
  )

  const handleFavoritePress = useCallback(async (): Promise<void> => {
    if (!isAuthenticated) {
      router.replace('/auth')
      return
    }

    if (isPending) return

    try {
      if (isFavorite) {
        await removeFavoriteMutation({ id, type })
      } else {
        await addFavoriteMutation({ id, type })
      }
    } catch (error) {
      Logger.breadcrumb('Failed to toggle favorite', 'error', error as Error)
    }
  }, [
    isAuthenticated,
    isPending,
    isFavorite,
    id,
    type,
    addFavoriteMutation,
    removeFavoriteMutation,
  ])

  return { handleFavoritePress, isFavorite, isPending }
}
