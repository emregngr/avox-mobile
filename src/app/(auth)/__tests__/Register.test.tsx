import { fireEvent, render, waitFor } from '@testing-library/react-native'
import { router, useLocalSearchParams } from 'expo-router'
import type { ReactNode } from 'react'
import React from 'react'
import { useForm } from 'react-hook-form'
import { SafeAreaProvider } from 'react-native-safe-area-context'

import Register from '@/app/(auth)/register'
import { useEmailRegister } from '@/hooks/services/useAuth'
import useFormStore from '@/store/form'
import useLocaleStore from '@/store/locale'

jest.mock('@/locales/i18next', () => ({
  getLocale: (key: string) => key,
}))

jest.mock('react-hook-form')

const mockedUseForm = useForm as jest.MockedFunction<typeof useForm>

jest.mock('@/store/locale')

const mockedUseLocaleStore = useLocaleStore as jest.MockedFunction<typeof useLocaleStore>

jest.mock('@/store/form')

const mockedUseFormStore = useFormStore as jest.MockedFunction<typeof useFormStore>

jest.mock('@/components/common', () => {
  const { View, Text, TouchableOpacity, TextInput } = require('react-native')

  return {
    Header: ({ title, backIconOnPress }: { title: string; backIconOnPress: () => void }) => (
      <>
        <Text testID="header-title">{title}</Text>
        <TouchableOpacity onPress={backIconOnPress} testID="back-button">
          <Text>Back</Text>
        </TouchableOpacity>
      </>
    ),

    SafeLayout: ({ children }: { children: ReactNode }) => (
      <View testID="safe-layout">{children}</View>
    ),

    TextInputField: ({ label, editable, ...props }: { label: string; editable: boolean }) => (
      <View>
        <Text>{label}</Text>
        <TextInput {...props} accessibilityLabel={label} editable={editable} />
      </View>
    ),

    CheckboxField: ({ name, onPressLabel }: { name: string; onPressLabel: () => void }) => (
      <TouchableOpacity onPress={onPressLabel} testID={`checkbox-${name}`} />
    ),

    ThemedGradientButton: ({
      onPress,
      label,
      disabled,
      loading,
    }: {
      onPress: () => void
      label: string
      disabled: boolean
      loading: boolean
    }) => (
      <TouchableOpacity disabled={disabled || loading} onPress={onPress} testID="register-button">
        <Text>{loading ? 'Loading...' : label}</Text>
      </TouchableOpacity>
    ),

    ThemedButtonText: ({
      onPress,
      label,
      disabled,
      loading,
    }: {
      onPress: () => void
      label: string
      disabled: boolean
      loading: boolean
    }) => (
      <TouchableOpacity
        disabled={disabled || loading}
        onPress={onPress}
        testID={`button-${label}`}
      />
    ),

    ThemedText: ({ children, ...props }: { children: string }) => (
      <Text {...props}>{children}</Text>
    ),
  }
})

const mockedReset = jest.fn()
const mockedlearForm = jest.fn()
const mockedSetFormValues = jest.fn()

const mockedFormValues = {
  firstName: 'John',
  lastName: 'Doe',
  email: 'john.doe@example.com',
  password: 'password123',
  privacyPolicy: true,
  termsOfUse: true,
}

jest.mock('@/hooks/services/useAuth')

const mockedUseEmailRegister = useEmailRegister as jest.MockedFunction<typeof useEmailRegister>

const mockedUseLocalSearchParams = useLocalSearchParams as jest.MockedFunction<
  typeof useLocalSearchParams
>

const renderWithSafeAreaProvider = (component: ReactNode) =>
  render(
    <SafeAreaProvider
      initialMetrics={{
        insets: { top: 0, left: 0, right: 0, bottom: 0 },
        frame: { x: 0, y: 0, width: 0, height: 0 },
      }}
    >
      {component}
    </SafeAreaProvider>,
  )

beforeEach(() => {
  mockedUseLocaleStore.mockReturnValue({ selectedLocale: 'en' })

  mockedUseFormStore.mockReturnValue({
    clearForm: mockedlearForm,
    formValues: {},
    setFormValues: mockedSetFormValues,
  })

  mockedUseEmailRegister.mockReturnValue({
    isPending: false,
    mutateAsync: mockedUseEmailRegister,
  } as any)

  mockedUseLocalSearchParams.mockReturnValue({})

  const mockGetValues = jest.fn((field?: keyof typeof mockedFormValues) => {
    if (field) {
      return mockedFormValues[field]
    }
    return mockedFormValues
  })

  mockedUseForm.mockReturnValue({
    control: {},
    handleSubmit: jest.fn(callback => () => callback(mockedFormValues)),
    formState: { errors: {}, isValid: true },
    getValues: mockGetValues,
    reset: mockedReset,
    setValue: jest.fn(),
    trigger: jest.fn(),
  } as any)
})

describe('Register Screen', () => {
  it('should render correctly', () => {
    const { getAllByText, getByLabelText, getByTestId } = renderWithSafeAreaProvider(<Register />)
    expect(getAllByText('register').length).toBeGreaterThan(0)
    expect(getByLabelText('firstName')).toBeTruthy()
    expect(getByLabelText('lastName')).toBeTruthy()
    expect(getByLabelText('email')).toBeTruthy()
    expect(getByLabelText('password')).toBeTruthy()
    expect(getByTestId('checkbox-privacyPolicy')).toBeTruthy()
    expect(getByTestId('checkbox-termsOfUse')).toBeTruthy()
    expect(getByTestId('register-button')).not.toBeDisabled()
  })

  it('should call register mutation on submit', async () => {
    const { getByTestId } = renderWithSafeAreaProvider(<Register />)
    const registerButton = getByTestId('register-button')
    fireEvent.press(registerButton)

    await waitFor(() => {
      expect(mockedUseEmailRegister).toHaveBeenCalledWith(mockedFormValues)
      expect(mockedlearForm).toHaveBeenCalled()
    })
  })

  it('should disable button when form is invalid after submission attempt', async () => {
    const mockGetValues = jest.fn().mockReturnValue(mockedFormValues)
    mockedUseForm.mockReturnValue({
      control: {},
      handleSubmit: jest.fn(callback => () => callback(mockedFormValues)),
      formState: { errors: {}, isValid: false },
      getValues: mockGetValues,
      reset: mockedReset,
      setValue: jest.fn(),
      trigger: jest.fn(),
    } as any)

    const { getByTestId } = renderWithSafeAreaProvider(<Register />)
    const registerButton = getByTestId('register-button')
    expect(registerButton).not.toBeDisabled()
    fireEvent.press(registerButton)
    await waitFor(() => {
      expect(registerButton).toBeDisabled()
    })
  })

  it('should show loading state and disable inputs', () => {
    ;(require('@/hooks/services/useAuth').useEmailRegister as jest.Mock).mockReturnValue({
      isPending: true,
      mutateAsync: mockedUseEmailRegister,
    })
    const { getByTestId, getByText, getByLabelText } = renderWithSafeAreaProvider(<Register />)
    expect(getByTestId('register-button')).toBeDisabled()
    expect(getByText('Loading...')).toBeTruthy()
    expect(getByLabelText('firstName').props.editable).toBe(false)
  })

  it('should navigate to login', () => {
    const { getByTestId } = renderWithSafeAreaProvider(<Register />)
    const loginButton = getByTestId('button-login')
    fireEvent.press(loginButton)
    expect(mockedReset).toHaveBeenCalled()
    expect(mockedlearForm).toHaveBeenCalled()
    expect(router.navigate).toHaveBeenCalledWith({
      params: { isRegisterParam: 'true' },
      pathname: '/login',
    })
  })

  it('should navigate to privacy policy', () => {
    const { getByTestId } = renderWithSafeAreaProvider(<Register />)
    const privacyPolicyCheckbox = getByTestId('checkbox-privacyPolicy')
    fireEvent.press(privacyPolicyCheckbox)
    expect(mockedSetFormValues).toHaveBeenCalledWith(mockedFormValues)
    expect(router.navigate).toHaveBeenCalledWith({
      params: {
        privacyPolicyParam: true,
        termsOfUseParam: true,
      },
      pathname: '/privacy-policy',
    })
  })

  it('should navigate to terms of use', () => {
    const { getByTestId } = renderWithSafeAreaProvider(<Register />)
    const termsOfUseCheckbox = getByTestId('checkbox-termsOfUse')
    fireEvent.press(termsOfUseCheckbox)
    expect(mockedSetFormValues).toHaveBeenCalledWith(mockedFormValues)
    expect(router.navigate).toHaveBeenCalledWith({
      params: {
        privacyPolicyParam: true,
        termsOfUseParam: true,
      },
      pathname: '/terms-of-use',
    })
  })

  it('should navigate back to auth screen by default', () => {
    const { getByTestId } = renderWithSafeAreaProvider(<Register />)
    const backButton = getByTestId('back-button')
    fireEvent.press(backButton)
    expect(mockedReset).toHaveBeenCalled()
    expect(mockedlearForm).toHaveBeenCalled()
    expect(router.replace).toHaveBeenCalledWith('/auth')
  })

  it('should navigate back to login screen if coming from there', () => {
    mockedUseLocalSearchParams.mockReturnValue({
      isLoginParam: 'true',
    })
    const { getByTestId } = renderWithSafeAreaProvider(<Register />)
    const backButton = getByTestId('back-button')
    fireEvent.press(backButton)
    expect(router.navigate).toHaveBeenCalledWith('/login')
  })
})

describe('Register Screen Snapshot', () => {
  it('should render the Register successfully', () => {
    const { toJSON } = renderWithSafeAreaProvider(<Register />)

    expect(toJSON()).toMatchSnapshot()
  })
})
