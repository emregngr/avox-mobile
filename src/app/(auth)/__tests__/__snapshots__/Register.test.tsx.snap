// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Register Screen Snapshot should render the Register successfully 1`] = `
<View
  testID="safe-layout"
>
  <Text
    testID="header-title"
  >
    register
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="back-button"
  >
    <Text>
      Back
    </Text>
  </View>
  <RCTScrollView
    bottomOffset={50}
    contentContainerClassName="px-4"
    contentContainerStyle={
      {
        "paddingBottom": 20,
        "paddingTop": 64,
      }
    }
    keyboardDismissMode="on-drag"
    keyboardShouldPersistTaps="handled"
    showsVerticalScrollIndicator={false}
  >
    <View>
      <View
        className="items-center my-8"
      >
        <View
          cachePolicy="memory-disk"
          contentFit="contain"
          source={
            {
              "default": "mocked-test-file",
            }
          }
          style={
            {
              "borderRadius": 50,
              "height": 100,
              "width": 100,
            }
          }
          testID="mocked-image"
          transition={0}
        />
      </View>
      <View
        className="gap-y-2"
      >
        <View>
          <Text>
            firstName
          </Text>
          <TextInput
            accessibilityLabel="firstName"
            control={{}}
            editable={true}
            keyboardType="default"
            name="firstName"
            onSubmitEditing={[Function]}
            placeholder="firstNamePlaceholder"
            returnKeyType="next"
            secureTextEntry={false}
            showToggle={false}
            testID="firstName"
          />
        </View>
        <View>
          <Text>
            lastName
          </Text>
          <TextInput
            accessibilityLabel="lastName"
            control={{}}
            editable={true}
            keyboardType="default"
            name="lastName"
            onSubmitEditing={[Function]}
            placeholder="lastNamePlaceholder"
            returnKeyType="next"
            secureTextEntry={false}
            showToggle={false}
            testID="lastName"
          />
        </View>
        <View>
          <Text>
            email
          </Text>
          <TextInput
            accessibilityLabel="email"
            control={{}}
            editable={true}
            keyboardType="email-address"
            name="email"
            onSubmitEditing={[Function]}
            placeholder="emailPlaceholder"
            returnKeyType="next"
            secureTextEntry={false}
            showToggle={false}
            testID="email"
          />
        </View>
        <View>
          <Text>
            password
          </Text>
          <TextInput
            accessibilityLabel="password"
            control={{}}
            editable={true}
            keyboardType="default"
            name="password"
            onSubmitEditing={[Function]}
            placeholder="passwordPlaceholder"
            returnKeyType="done"
            secureTextEntry={true}
            showToggle={true}
            testID="password"
          />
        </View>
      </View>
      <View
        className="my-6 gap-y-2"
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="checkbox-privacyPolicy"
        />
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="checkbox-termsOfUse"
        />
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="register-button"
      >
        <Text>
          register
        </Text>
      </View>
      <View
        className="flex-row justify-center items-center mt-6"
      >
        <Text
          color="text-70"
          type="body2"
        >
          alreadyHaveAccount
        </Text>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="button-login"
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
